nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PublishSlack.java

```diff
@@ -0,0 +1,516 @@
+import org.apache.nifi.expression.AttributeExpression;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.flowfile.attributes.CoreAttributes;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.processors.slack.consume.ConsumeSlackUtil;
+import org.apache.nifi.processors.slack.publish.ContentMessageStrategy;
+import org.apache.nifi.processors.slack.publish.MessageOnlyStrategy;
+import org.apache.nifi.processors.slack.publish.PublishConfig;
+import org.apache.nifi.processors.slack.publish.PublishSlackClient;
+import org.apache.nifi.processors.slack.publish.PublishStrategy;
+import org.apache.nifi.processors.slack.publish.UploadAttachmentStrategy;
+import org.apache.nifi.processors.slack.publish.UploadLinkStrategy;
+import org.apache.nifi.processors.slack.publish.UploadOnlyStrategy;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import static org.apache.nifi.util.StringUtils.isEmpty;
+
+
+@Tags({"slack", "publish", "notify", "upload", "message"})
+@CapabilityDescription("Uploads the FlowFile content (e.g. an image) as a file, and/or Sends a message on Slack. The FlowFile content can be used as the message text or attached to the message.")
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@DynamicProperty(name = "<Arbitrary name>", value = "JSON snippet specifying a Slack message \"attachment\"",
+        description = "The property value will be converted to JSON and will be added to the array of attachments in the JSON payload being sent to Slack." +
+                " The property name will not be used by the processor.",
+        expressionLanguageScope = ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+@WritesAttribute(attribute="slack.file.url", description = "The Slack URL of the uploaded file. It will be added if 'Upload FlowFile' has been set to 'Yes'.")
+public class PublishSlack extends AbstractProcessor {
+
+    private static final String UPLOAD_ONLY = "upload-only";
+    private static final String MESSAGE_ONLY = "message-only";
+    private static final String CONTENT_MESSAGE = "content-message";
+    private static final String UPLOAD_ATTACH = "upload-attach";
+    private static final String UPLOAD_LINK = "upload-link";
+
+    public static final PropertyDescriptor BOT_TOKEN = new PropertyDescriptor.Builder()
+            .name("Bot Token")
+            .description("The Bot Token that is registered to your Slack application")
```

Comment:
It'd be very useful if the required scopes would be listed here as it is done e.g. for ConsumeSlack ACCESS_TOKEN property.